common/blockchain/src/ibc_state/proof.rs

```diff
@@ -0,0 +1,435 @@
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+
+use lib::hash::CryptoHash;
+
+mod ibc {
+    pub use ibc_core_commitment_types::commitment::{
+        CommitmentPrefix, CommitmentProofBytes, CommitmentRoot,
+    };
+    pub use ibc_core_host::types::path;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct IbcProof {
+    /// Serialised proof.
+    pub proof: ibc::CommitmentProofBytes,
+    /// Commitment root.
+    pub root: ibc::CommitmentRoot,
+    /// Value stored at the path (if it exists).
+    pub value: Option<CryptoHash>,
+}
+
+impl IbcProof {
+    /// Returns commitment prefix to use during verification.
+    pub fn prefix(&self) -> ibc::CommitmentPrefix { Default::default() }
+}
+
+
+#[derive(Clone, Debug, PartialEq, Eq, derive_more::From)]
+pub enum GenerateError {
+    /// State root in block header and root of trie don’t match.
+    WrongState,
+
+    /// Error reading data from the trie.
+    BadTrie(sealable_trie::Error),
+
+    /// Invalid path.
+    BadPath(trie_ids::path_info::Error),
+}
+
+/// Generates a proof for given path.
+///
+/// `block_header` is header whose hash will be the commitment root.  It’s
+/// state root must correspond to `trie`’s root.  `path` specifies IBC path
+/// of the value that needs proof.
+///
+/// # Proof format
+///
+/// In most cases, proof is Borsh-serialised `(blockchain::BlockHeader,
+/// sealable_trie::proof::Proof)` pair.  The header at the front is
+/// necessary to determine state root (recall that `root` is the block hash
+/// and not state root).
+///
+/// However, if `path` is one of `SeqSend`, `SeqRecv` or `SeqAck` than proof
+/// further contain two big-endian encoded `u64` numbers holding the other
+/// two sequence numbers.
+///
+/// For example, if `path` is `SeqRecv`, the `proof` must at the end include
+/// send sequence number and ack sequence number.  For example, if next send
+/// sequence is `7`, next ack sequence is `5` and path is `SeqRecv` the
+/// proof will end with `be(7) || be(5)` (where `be` denotes encoding 64-bit
+/// number as big endian).
+///
+/// This addition is necessary because sequence numbers are stored together
+/// within a single trie value.  For example, proving the next receive
+/// sequence is `4` requires proving `be(7), be(4), be(5), be(0)].  For
+/// verifier to know what value it checks, it needs to be provided all of
+/// the sequence numbers.
+///
+/// (Note that Borsh uses little endian to encode integers so the sequence
+/// numbers cannot be simply borsh deserialised.)
+pub fn generate<A: sealable_trie::Allocator>(
```

Comment:
why not `generate_proof`

Reply:
Well, the full path to to the function is `ibc_state::proof::generate` so `proof` is already there.